### PR TITLE
Dashboard filters unsaved query params

### PR DESCRIPTION
**Pull request by mattgraham1995**
## Description
It is not easy to share a dashboard with multiple different filter configurations. Currently, you need to save multiple copies and send out links that way.

## Proposed Changes

- Tie a filters query param to the dashboard's filter fragments property
- Updating a dashboards filters now updates the query params, then the route reloads and the model hook will update the dashboard model from those query params
## Screenshots
Dashboard-filter-query-params

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2019, Yahoo Holdings Inc.
+ * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
+ */
+
+import Controller from '@ember/controller';
+import { inject as service } from '@ember/service';
+import { isArray } from '@ember/array';
```
Comment (jkusa) [inline packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: Can we use this instead: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+  /**
+   * @property {String[]} queryParams
+   */
+  queryParams: ['filters'],
+
+  /**
+   * Update Model to have dirty attributes that match the filter query params
+   *
+   * @method addFiltersFromQueryParams
```
Comment (jkusa) [inline packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: is this private?

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  /**
+   * @private
+   * @method _removeAllFiltersFromDashboard
+   * @param {Object} dashboard
+   */
+  _removeAllFiltersFromDashboard(dashboard) {
+    const filters = dashboard.filters;
+    while (filters.get('firstObject')) {
+      filters.popObject();
+    }
```
Comment (jkusa) [inline packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: Is this: `dashboard.filters.length = 0`?

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+      this._removeAllFiltersFromDashboard(dashboard);
+
+      await Promise.all(
+        decompressedFilters.filters.map(async fil => {
+          const newFragmentFields = {
+            field: fil.field,
+            operator: fil.operator,
+            rawValues: fil.values,
+            dimension: this.metadataService.getById('dimension', fil.dimension)
```
Comment (daniellevinson) [inline packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: should this be `this.get('metadataService')` for Ember 2.x?

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+          };
+
+          const newFragment = this.store.createFragment('bard-request/fragments/filter', newFragmentFields);
```
Comment (daniellevinson) [inline packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: same for `store` and `compression` services

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 6]
```diff
@@ -91,4 +91,170 @@ module('Acceptance | Dashboard Filters', function(hooks) {
...
+    assert.equal(currentURL(), '/dashboards/1/view', 'No query params are set');
+
+    assert.equal(dataRequests.length, 3, 'Three requests have been run since route entry');
+
+    await click('.dashboard-filters__expand-button');
+    await click('.dashboard-filters-expanded__add-filter-button');
+
+    await selectChoose('.dashboard-dimension-selector', 'Property');
+
+    assert.equal(
```
Comment (kevinhinterlong) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: Can we assert on the decompressed filters?

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 7]
```diff
@@ -91,4 +91,170 @@ module('Acceptance | Dashboard Filters', function(hooks) {
...
+      '/dashboards/1/view?filters=EQbwOsBmCWA2AuBTATgZwgLgNrmAE2gFtEA7VaAexMwgAdkLaV4BPCAGgkZQEN4LkNYNGrBOwAG49YAV0Tpg2ALriYiWHiHRNwAL5LdwIAA',
+      'Query params are added on filter add'
+    );
+    assert.equal(dataRequests.length, 3, 'No new requests run on filter add (model hook should use cached data)');
+
+    await fillIn('.filter-builder-dimension__values input', '1');
+    await selectChoose('.filter-builder-dimension__values', '.item-row', 0);
+
+    assert.equal(dataRequests.length, 6, 'New requests are sent once filter values are set');
+    assert.equal(
```
Comment (kevinhinterlong) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: Here too

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  }
+});
+
+/**
+ * Move values from `rawValues` property to `values` property
+ * @function cookValues
+ * @param {Object} obj - object with rawValues property
+ * @returns {Object} object with values property and no rawValues property
+ */
+function cookValues(obj) {
```
Comment (cythrawll) [inline packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: nice

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js [hunk 9]
```diff
@@ -13,18 +15,114 @@ export default Route.extend({
...
+
+    const originalFilters = dashboard.get('filters').serialize();
+    await this._addFiltersFromQueryParams(dashboard, filterQueryParams);
+    const newFilters = dashboard.get('filters').serialize();
+
+    //If we just added a new filter (no values yet) return the last set of widget data because the requests will not have changed
+    if (
+      cachedWidgetData &&
+      newFilters.length === originalFilters.length + 1 &&
+      newFilters.any(fil => fil.values.length === 0)
```
[the remark this answers is not in the document]
Comment (kevinhinterlong) [reply packages/dashboards/addon/routes/dashboards/dashboard/view.js]:
```suggestion
      newFilters.filter(fil => fil.values.length === 0).length === originalFilters.filter(fil => fil.values.length === 0).length
```

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js (hunk 8 above)
Comment (mattgraham1995) [reply packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: 😎

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mattgraham1995) [reply packages/dashboards/addon/routes/dashboards/dashboard/view.js]: Okay, this case is explicitly tested now and I've added the case where we remove an empty filter as well.

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js [hunk 10]
```diff
@@ -13,18 +15,132 @@ export default Route.extend({
...
+    const filterQueryParams = get(transition, 'queryParams.filters');
+    const cachedWidgetData = this.get('_widgetDataCache');
+
+    const originalFilters = dashboard.get('filters').serialize();
+    const originalValuelessFilters = originalFilters.filter(fil => fil.values.length === 0);
+    await this._addFiltersFromQueryParams(dashboard, filterQueryParams);
+    const newFilters = dashboard.get('filters').serialize();
+    const newValuelessFilters = newFilters.filter(fil => fil.values.length === 0);
+
+    const wasNewFilterAdded =
```
Comment (kevinhinterlong) [inline packages/dashboards/addon/routes/dashboards/dashboard/view.js]: Maybe `wasNewEmptyFilterAdded`

File: packages/dashboards/addon/controllers/dashboards/dashboard/view.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+     * @param {Object} originalFilter
+     * @param {Object} changeSet
+     */
+    async updateFilter(dashboard, originalFilter, changeSet) {
+      const origFilter = originalFilter.serialize();
+      const newFilters = get(dashboard, 'filters')
+        .toArray()
+        .map(fil => fil.serialize()); //Native array of serialized filters
+      const newFilter = newFilters.find(fil => isEqual(fil, origFilter));
+      const validChangeSet = cookValues(changeSet);
```
Comment (jkusa) [inline packages/dashboards/addon/controllers/dashboards/dashboard/view.js]: Could we hydrate a new fragment and then call `.serialize()`?

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js [hunk 12]
```diff
@@ -13,18 +15,146 @@ export default Route.extend({
...
+    const originalFilters = dashboard.get('filters').serialize();
+    try {
+      await this._addFiltersFromQueryParams(dashboard, filterQueryParams);
+    } catch (e) {
+      get(this, 'naviNotifications').add({
+        message: `Error decoding filter query params. Using default dashboard filters`,
+        type: 'danger',
+        timeout: 'medium'
       });
+      throw e;
```
Comment (mattgraham1995) [inline packages/dashboards/addon/routes/dashboards/dashboard/view.js]: console.error

Comment (mattgraham1995) [reply packages/dashboards/addon/routes/dashboards/dashboard/view.js]: or not

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js [hunk 13]
```diff
@@ -13,18 +15,146 @@ export default Route.extend({
...
+
+    /**
+     * If we just added a new filter (no values yet) or just removed a filter with no values,
+     * return the last set of widget data because the requests will not have changed because
+     * empty filters are pruned from the request
+     */
+    if (cachedWidgetData && (wasEmptyFilterAdded || wasEmptyFilterRemoved)) {
+      return { dashboard, dataForWidget: cachedWidgetData };
+    }
+    const dataForWidget = await this.get('dashboardData').fetchDataForDashboard(dashboard);
```
Comment (jkusa) [inline packages/dashboards/addon/routes/dashboards/dashboard/view.js]: Caching down a layer to only re-fetch widget request that changed?

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js [hunk 14]
```diff
@@ -13,18 +15,146 @@ export default Route.extend({
...
+   * @override
+   * @method resetController - reset query params on exit of route
+   * @param {Object} controller
+   * @param {Boolean} isExiting
+   */
+  resetController(controller, isExiting /*transition*/) {
+    this._super(...arguments);
+
+    if (isExiting) {
+      controller.resetModel();
```
Comment (mattgraham1995) [inline packages/dashboards/addon/routes/dashboards/dashboard/view.js]: clear cache

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js [hunk 15]
```diff
@@ -13,18 +15,146 @@ export default Route.extend({
...
+
+      if (!decompressedFilters.hasOwnProperty('filters') || !Array.isArray(decompressedFilters.filters)) {
+        throw Error('Filter query params are not valid filters');
+      }
+
+      //Remove all filters from the fragment array
+      modelFilters.length = 0;
+
+      await Promise.all(
+        decompressedFilters.filters.map(async fil => {
```
Comment (jkusa) [inline packages/dashboards/addon/routes/dashboards/dashboard/view.js]: do we need `async fil`?

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js (hunk 14 above)
Comment (mattgraham1995) [reply packages/dashboards/addon/routes/dashboards/dashboard/view.js]: Move to deactivate

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 16]
```diff
@@ -91,4 +97,284 @@ module('Acceptance | Dashboard Filters', function(hooks) {
...
+    await click('.navi-dashboard__breadcrumb-link');
+    assert.equal(currentURL(), '/dashboards', 'Routed back to dashboards route with no active query params');
+    await click('.navi-collection__row1 a');
+    assert.equal(
+      currentURL(),
+      '/dashboards/2/view',
+      'Returning to index view and back to the dashboard clears the dirty state and query params'
+    );
+    assert.deepEqual(extractCollapsedFilters(), INITIAL_FILTERS, 'Initial filters are present');
+    assert.dom('.navi-dashboard__save-dialogue').isNotVisible('The dashboard is in a clean state');
```
Comment (mattgraham1995) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: Test that requests are formatted correctly

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 17]
```diff
@@ -91,4 +97,284 @@ module('Acceptance | Dashboard Filters', function(hooks) {
       'Correct filters are saved with the dashboard'
     );
   });
+
+  test('dashboard filter query params - model changes query params', async function(assert) {
```
Comment (mattgraham1995) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: rename test

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 18]
```diff
@@ -91,4 +97,284 @@ module('Acceptance | Dashboard Filters', function(hooks) {
...
+
+  test('dashboard filter query params - returns cached dashboard widget data on filter add', async function(assert) {
+    assert.expect(9);
+
+    let dataRequests = [];
+    server.urlPrefix = `${config.navi.dataSources[0].uri}/v1`;
+    server.pretender.handledRequest = (verb, url, req) => {
+      if (verb == 'GET' && url.includes('/v1/data')) {
+        dataRequests.push(req);
+      }
```
Comment (mattgraham1995) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: return { rows: [] };
remove GET check

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 19]
```diff
@@ -91,4 +97,284 @@ module('Acceptance | Dashboard Filters', function(hooks) {
...
+    /**
+     * This is to test a specific corner case where a new filter is added (no values) and then the user pastes a
+     * link to the same dashboard with the same filters except has one more added filter with a value
+     */
+    assert.expect(6);
+
+    let dataRequests = [];
+    server.urlPrefix = `${config.navi.dataSources[0].uri}/v1`;
+    server.pretender.handledRequest = (verb, url, req) => {
+      if (verb == 'GET' && url.includes('/v1/data')) {
```
Comment (mattgraham1995) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: here too

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 20]
```diff
@@ -91,4 +97,284 @@ module('Acceptance | Dashboard Filters', function(hooks) {
...
+          }
+        ]
+      },
+      'A filter with no values was added'
+    );
+    assert.equal(dataRequests.length, 3, 'No new requests run on filter add (model hook should use cached data)');
+
+    await visit(
+      '/dashboards/1/view?filters=EQbwOsBmCWA2AuBTATgZwgLgNrmAE2gFtEA7VaAexMwgAdkLaV4BPCAGgkZQEN4LkNYNGrBOwAG49YAV0Tpg2ALriYiWHiHRNwAL7tcBYmUqiMEHgHNEHLk2R8BW0eKmz5mLBAC0AZggqEGoaWjq6SrrAQA',
+      { refresh: true }
```
Comment (mattgraham1995) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: remove

File: packages/dashboards/tests/integration/components/dashboard-filters-expanded-test.js [hunk 21]
```diff
@@ -55,7 +59,22 @@ module('Integration | Component | dashboard filters expanded', function(hooks) {
         }
       ]
     };
-    await render(hbs`{{dashboard-filters-expanded dashboard=dashboard}}`);
+    this.onUpdateFilter = function() {
```
Comment (jkusa) [inline packages/dashboards/tests/integration/components/dashboard-filters-expanded-test.js]: `this.onUpdateFilter = () => null`

File: packages/dashboards/tests/acceptance/dashboard-filter-test.js [hunk 22]
```diff
@@ -91,4 +97,284 @@ module('Acceptance | Dashboard Filters', function(hooks) {
...
+            values: ['-3']
+          }
+        ]
+      },
+      'The expected filters are set in the query params'
+    );
+  });
 });
+
+/**
```
Comment (jkusa) [inline packages/dashboards/tests/acceptance/dashboard-filter-test.js]: Can we add a negative test for query params?

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js (hunk 13 above)
Comment (mattgraham1995) [reply packages/dashboards/addon/routes/dashboards/dashboard/view.js]: Will be done in a later PR as an optimization

File: packages/dashboards/addon/routes/dashboards/dashboard/view.js [hunk 23]
```diff
@@ -150,11 +147,11 @@ export default Route.extend({
...
    * @param {Boolean} isExiting
    */
-  resetController(controller, isExiting /*transition*/) {
+  deactivate() {
     this._super(...arguments);
 
-    if (isExiting) {
-      controller.resetModel();
-    }
+    this.controllerFor('dashboards.dashboard.view').set('filters', null);
```
Comment (jkusa) [inline packages/dashboards/addon/routes/dashboards/dashboard/view.js]: Can this work
`this.controller`

 https://api.emberjs.com/ember/3.12/classes/Route/properties/controller?anchor=controller
